### ISSUE TITLE
Issue with sas3ircu resource

### ISSUE THREAD

**Issue by dashmage** [canonical/hardware-observer-operator#55]
```sh
juju deploy ubuntu --series jammy
juju deploy hardware-observer --channel edge
juju relate ubuntu hardware-observer

juju deploy grafana-agent
juju relate ubuntu grafana-agent
juju relate hardware-observer grafana-agent
```

After setting up zookeeper, hardware-observer and grafana-agent on a machine with `sas3ircu` compatibility and attaching the `sas3ircu` resource, I'm not able to see the sas3ircu metrics on Prometheus.

The `sas3ircu` command is registered as successful while checking the exporter metrics. The `lsi_sas_3_xyz` metrics are also visible while running `curl localhost:10000`.

```
ubuntu@bomberto:~$ curl localhost:10000
(...)
# HELP lsi_sas_3_controllers Number of LSI SAS-{self.version} controllers
# TYPE lsi_sas_3_controllers gauge
lsi_sas_3_controllers 1.0
# HELP sas3ircu_command_success Indicates if the command is successful or not
# TYPE sas3ircu_command_success gauge
sas3ircu_command_success 1.0
# HELP lsi_sas_3_ir_volumes Number of IR volumes
# TYPE lsi_sas_3_ir_volumes gauge
lsi_sas_3_ir_volumes{controller_id="0"} 0.0
# HELP lsi_sas_3_ready_ir_volumes Number of ready IR volumes
# TYPE lsi_sas_3_ready_ir_volumes gauge
lsi_sas_3_ready_ir_volumes{controller_id="0"} 0.0
# HELP lsi_sas_3_unready_ir_volumes Number of unready IR volumes
# TYPE lsi_sas_3_unready_ir_volumes gauge
lsi_sas_3_unready_ir_volumes{controller_id="0"} 0.0
# HELP lsi_sas_3_physical_devices Number of physical devices
# TYPE lsi_sas_3_physical_devices gauge
lsi_sas_3_physical_devices{controller_id="0"} 2.0
# HELP lsi_sas_3_ready_physical_devices Number of ready physical devices
# TYPE lsi_sas_3_ready_physical_devices gauge
lsi_sas_3_ready_physical_devices{controller_id="0"} 2.0
# HELP lsi_sas_3_unready_physical_devices Number of unready physical devices
# TYPE lsi_sas_3_unready_physical_devices gauge
lsi_sas_3_unready_physical_devices{controller_id="0"} 0.0
# HELP lsi_sas_3_enclosure_info Show the information about the enclosure
# TYPE lsi_sas_3_enclosure_info gauge
lsi_sas_3_enclosure_info{controller_id="0",enclosure_id="1",num_slots="10",start_slot="0"} 1.0
```

**Comment by dashmage** [comment]
I'm intermittently able to view the `sas3ircu` metrics on Prometheus now.

![image](https://github.com/canonical/hardware-observer-operator/assets/25390807/be188f98-7190-486c-968f-959422e342c5)

![image](https://github.com/canonical/hardware-observer-operator/assets/25390807/f215ba31-eb6a-4ac4-b613-e196a24d44b2)

Not sure why there's periods of time where the metrics are not received.

**Comment by dashmage** [comment]
After some troubleshooting it was found that the reason that there are gaps in the graph because querying the exporter endpoint is frequently taking more than 10s which is the default scrape timeout configured on Prometheus. This is because the sensor metrics exported from redfish (which is also present on this server) have a lot of information which is taking time to scrape and send to Prometheus. 

There is also another issue where if the credentials for redfish are incorrect, the hardware-exporter service keeps trying to connect to redfish and takes time. This also causes to exceed the scrape timeout. This can be solved by setting timeout and max_retries values while calling the [redfish_client](https://github.com/DMTF/python-redfish-library/blob/main/src/redfish/rest/v1.py#L1125) method to create the redfish object.

Running `curl localhost:10000` can take more than 10s which is seen [here](https://pastebin.ubuntu.com/p/CnthSFs4b3/). We can also see from the same link that there are a lot of items returned from the redfish sensor metric.

One approach to solve this problem is by adding a cache mechanism which the exporter can query instead of directly fetching it from the redfish API each time.

**Comment by dashmage** [comment]
We can't cache the results of the `_get_sensor_data` method in `collectors/redfish.py` using something like `@lru_cache`. This is because even with the same arguments for the method (redfish host, user, password) the values returned by the sensors could differ and we wouldn't want them to be cached.

I'm not sure if we can cache the sensor values returned by the `redfish_utilities.get_sensors(redfish_obj)` itself since:
1. It is only after executing the method call that we know whether something has changed or not. The arguments (`redfish_obj`) stays the same.
2. These sensor values like temperature, voltage or fan speed for example would be changing constantly which make it hard to cache.

**Comment by dashmage** [comment]
Running the `redfish_utilities.get_sensors()` method manually returns all the sensor details in <1s.

```sh
ubuntu@bomberto:~$ time python3 redfish_test.py 
[{'ChassisName': '1',
  'Readings': [{'Health': 'OK',
                'LowerCaution': None,
                (...)
                'Units': '%',
                'UpperCaution': None,
                'UpperCritical': None,
                'UpperFatal': None}]}]

real	0m0.672s
user	0m0.109s
sys	0m0.004s
```

```python
# cat redfish_test.py

import redfish_utilities as rf
import redfish
from pprint import pprint as pp

host = "https://1.2.3.4"
user = "myuser"
pwd = "secret"

redfish_obj = redfish.redfish_client(base_url=host, username=user, password=pwd)
redfish_obj.login(auth="session")

sensor_data = rf.get_sensors(redfish_obj)
pp(sensor_data)
```

**Comment by jneo8** [comment]
I would like to discuss about the cache issue on https://github.com/canonical/prometheus-hardware-exporter/issues/33 .
There are not too much we can do here.

**Comment by jneo8** [comment]
> So, we could look into increasing the scrape_timeout config for Prometheus itself somehow.

It's put the cart before the horse.

> Running the redfish_utilities.get_sensors() method manually returns all the sensor details in <1s.

I believe we missing a lot of things in your script. Also we still need to prevent the wrong user/password situation.


Also can we move the discuss to https://github.com/canonical/prometheus-hardware-exporter/issues/33 ?

**Comment by dashmage** [comment]
PR is merged! 
https://github.com/canonical/prometheus-hardware-exporter/pull/34
